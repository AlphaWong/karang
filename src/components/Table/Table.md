#### Usage:

```js static
import { Table } from 'lalamove-ui';

// <Table columns={[object]} data={[object]}>
```

**Basic**
```js
<Table
  columns={[
    { key: 'title', label: 'Title' },
    { key: 'rating', label: 'How good' },
    { key: 'year', label: 'Year' },
  ]}
  data={[
    {
      id: 'breaking-bad',
      title: 'Breaking Bad',
      rating: 5,
      year: '2008',
    },
    {
      id: 'better-call-saul',
      title: 'Better Call Saul',
      rating: 5,
      year: '2016',

    },
    {
      id: 'the-office',
      title: 'The Office',
      rating: 3,
      year: '2004',

    }
  ]}
/>
```

**Render function**

Use `render` function to customize how you wish to render the table cell. 
Here the rating value is rendered as a `<Rating />` component.

```js static
columns={[
  { key: 'title', label: 'Title' },
  {
    key: 'rating',
    label: 'How good',
    render: val => <Rating value={val} />
  },
  { key: 'year', label: 'Year' },
]}
```

```js
<Table
  columns={[
    { key: 't', label: 'Title' },
    {
      key: 'r',
      label: 'How good',
      render: val => <Rating value={val} />
    },
    { key: 'y', label: 'Year' },
  ]}
  data={[
    { id: 'breaking-bad', t: 'Breaking Bad', r: 5, y: '2008' },
    { id: 'better-call-saul', t: 'Better Call Saul', r: 5, y: '2016' },
    { id: 'the-office', t: 'The Office', r: 3, y: '2004' }
  ]}
/>
```

**Sortable columns**

_Supports frontend and api sorting_

Table data is sortable by column. You can choose to have it sorted purely presentationally, or sort `props.data` yourself when `onSort` function is triggered.

```js
<Table
  hoverable
  columns={[
    {
      key: 'name',
      label: 'Name',
      onSort: (key, order) => (a, b) => {
        switch (order) {
          case 'desc':
            return b[key].localeCompare(a[key]);
          case 'asc':
            return a[key].localeCompare(b[key]);
          default:
            return 0;
        }
      },
    },
    {
      key: 'age',
      label: 'Years on Earth',
      onSort: (key, order) => {
        return (a, b) => {
          switch (order) {
            case 'desc':
              return b[key] - a[key];
            case 'asc':
              return a[key] - b[key];
            default:
              return 0;
          }
        };
      },
    },
    {
      key: 'job',
      label: 'Occupation (sort yourself, see console)',
      // colKey == hobbies
      // order == one of 'default', 'desc', 'asc'
      onSort: (colKey, order) => {
        /* May trigger api with sorting data here */
        console.log(colKey, order);
      },
    },
    {
      key: 'hobbies',
      label: 'Fun',
      onSort: (key, order) => (a, b) => {
        switch (order) {
          case 'desc':
            return b[key].localeCompare(a[key]);
          case 'asc':
            return a[key].localeCompare(b[key]);
          default:
            return 0;
        }
      },
    },
  ]}
  data={[
    { id: 24523, name: 'Wong', age: 43, job: 'Business man', hobbies: '購買' },
    { id: 47657, name: 'Smith', age: 35, job: 'Janitor', hobbies: '吃' },
    { id: 5675, name: 'Trump', age: 72, job: 'PUTUS', hobbies: '坐' },
    { id: 123, name: 'Chan', age: 54, job: 'CEO', hobbies: '籃球' },
    { id: 352, name: 'Mannings', age: 31, job: 'Carpenter', hobbies: '足球' },
    { id: 312, name: 'Chow', age: 12, job: 'Hobo', hobbies: '曲棍球' },
  ]}
/>
```

**Derived column**

_Derived data from other columns, and put it in it's own column_

We can create new column and fill column cells with anything. This is how the table will normally render the following data.

```js static
const data = [
  { id: 24523, name: 'Wong', age: 43, income: 40000 },
  { id: 47657, name: 'Smith', age: 35, income: 20000 },
  { id: 5675, name: 'Trump', age: 72, income: 1 },
  { id: 123, name: 'Chan', age: 54, income: 30000 },
  { id: 352, name: 'Mannings', age: 31, income: 80000 },
  { id: 312, name: 'Chow', age: 12, income: 0 },
];

```

```js
<Table
  columns={[
    { key: 'name', label: 'Name' },
    { key: 'age', label: 'Years on Earth' },
    { key: 'income', label: '💵 Money' },
  ]}
  data={[
    { id: 24523, name: 'Wong', age: 43, income: 40000 },
    { id: 47657, name: 'Smith', age: 35, income: 20000 },
    { id: 5675, name: 'Trump', age: 72, income: 1 },
    { id: 123, name: 'Chan', age: 54, income: 30000 },
    { id: 352, name: 'Mannings', age: 31, income: 80000 },
    { id: 312, name: 'Chow', age: 12, income: 0 },
  ]}
/>
```

We can derive data from other columns. As we all know handsome-ness is based on an individual's income and age. We can create a new column named 'Handsome-ness'. 

Notice it does not have a corresponding data property, by using the `render` function, we can conjure up our own data based on `income` and `age` property of each individual record comparing to the average.

```js static
columns={[
  ...,
  {
    key: 'rating',
    label: 'Handsome-ness',
    render(_, dude, dudes) {
      const maxAge = Math.max(...dudes.map(({ age }) => age));
      const maxIncome = Math.max(...dudes.map(({ income }) => income));
      const ageRate = 6 - Math.round((dude.age / maxAge) * 5);
      const incomeRate = Math.round((dude.income / maxIncome) * 5);
      const handsomeNess = Math.round((ageRate + incomeRate) / 2) || 1;
      return <Rating value={handsomeNess} />;
    },
  },
]}
```

```js
<Table
  columns={[
    { key: 'name', label: 'Name' },
    { key: 'age', label: 'Years on Earth' },
    { key: 'income', label: '💵 Money' },
    {
      key: 'rating',
      label: 'Handsome-ness',
      render(_, dude, dudes) {
        const maxAge = Math.max(...dudes.map(({ age }) => age));
        const maxIncome = Math.max(...dudes.map(({ income }) => income));
        const ageRate = 6 - Math.round((dude.age / maxAge) * 5);
        const incomeRate = Math.round((dude.income / maxIncome) * 5);
        const rate = Math.round((ageRate + incomeRate) / 2);
        return <Rating value={rate || 1} />;
      },
    },
  ]}
  data={[
    { id: 24523, name: 'Wong', age: 43, income: 40000 },
    { id: 47657, name: 'Smith', age: 35, income: 20000 },
    { id: 5675, name: 'Trump', age: 72, income: 1 },
    { id: 123, name: 'Chan', age: 54, income: 30000 },
    { id: 352, name: 'Mannings', age: 31, income: 80000 },
    { id: 312, name: 'Chow', age: 12, income: 0 },
  ]}
/>
```

**Usage with `<Pagination/>`**
