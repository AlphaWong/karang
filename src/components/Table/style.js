import styled, { css } from 'styled-components';
import { lighten } from 'polished';
import { fontSize, fontWeight } from 'styles/fonts';
import { offWhite, orange, silver } from 'styles/colors';

export const SCTable = styled.table`
  width: 100%;
  border-spacing: 0;
  line-height: 1.4;

  thead th {
    padding: 0.5em 1em;
    border-top: 1px ${offWhite} solid;
    border-bottom: 1px ${offWhite} solid;
    color: ${silver};
    font-size: ${fontSize.small}px;
    font-weight: ${fontWeight.bold};
    text-align: left;
  }

  tbody td {
    padding: 1em;
    vertical-align: top;

    &:first-of-type {
      border-left: 4px solid transparent;
    }
  }
`;

export const Row = styled.tr`
  &:nth-child(even) {
    background: ${lighten(0.08, offWhite)};
  }
  ${({ hoverable }) =>
    hoverable &&
    css`
      &:hover {
        background: ${lighten(0.05, offWhite)};

        > td:first-of-type {
          border-left-color: ${orange};
        }
      }
    `};
`;
