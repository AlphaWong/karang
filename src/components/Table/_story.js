import React from 'react';
import { storiesOf } from '@storybook/react';
import { boolean } from '@storybook/addon-knobs/react';

import Table from './index';

const columns = [
  {
    key: 'name',
    label: 'Name',
    onSort: key => {
      console.log('look', key);
    },
  },
  {
    key: 'job',
    label: 'Occupation',
  },
  {
    key: 'hobbies',
    label: 'Fun',
  },
];

const data = [
  {
    id: 123,
    name: 'Chan',
    job: 'CEO',
    hobbies: 'basketball',
  },
  {
    id: 312,
    name: 'Chow',
    job: 'Hobo',
    hobbies: 'hockey',
  },
  {
    id: 352,
    name: 'Mannings',
    job: 'Carpenter',
    hobbies: 'football',
  },
  {
    id: 24523,
    name: 'Wong',
    job: 'Business man',
    hobbies: 'buy',
  },
  {
    id: 47657,
    name: 'Smith',
    job: 'Janitor',
    hobbies: 'eat',
  },
  {
    id: 5675,
    name: 'Trump',
    job: 'PUTUS',
    hobbies: 'sit',
  },
];

storiesOf('Table', module).add('Basic', () => (
  <Table
    hoverable={boolean('hoverable', false)}
    columns={columns}
    data={data}
  />
));
