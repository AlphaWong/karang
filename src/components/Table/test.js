/* global mount */
import React from 'react';
import Table from './index';

const mockColumns = [
  {
    key: 'name',
    label: 'Name',
  },
  {
    key: 'job',
    label: 'Occupation',
  },
  {
    key: 'hobbies',
    label: 'Fun',
  },
];

const tableData = [
  {
    id: 123,
    name: 'Chan',
    job: 'CEO',
    hobbies: 'basketball',
  },
  {
    id: 312,
    name: 'Chow',
    job: 'Hobo',
    hobbies: 'hockey',
  },
  {
    id: 352,
    name: 'Mannings',
    job: 'Carpenter',
    hobbies: 'football',
  },
  {
    id: 24523,
    name: 'Wong',
    job: 'Business man',
    hobbies: 'buy',
  },
  {
    id: 47657,
    name: 'Smith',
    job: 'Janitor',
    hobbies: 'eat',
  },
  {
    id: 5675,
    name: 'Trump',
    job: 'PUTUS',
    hobbies: 'sit',
  },
];

test('not crash', () => {
  expect(() => <Table columns={mockColumns} data={tableData} />).not.toThrow();
});

test('snapshots', () => {
  expect(
    mount(<Table columns={mockColumns} data={tableData} />)
  ).toMatchSnapshot();
});
