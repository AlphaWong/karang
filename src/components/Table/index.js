import React, { Component } from 'react';
import { bool, string, shape, arrayOf, func, object } from 'prop-types';

import { SCTable, Row } from './style';

function findColumn(key, listOfcols) {
  return listOfcols.reduce((found, col) => {
    if (found === null && col.key === key) {
      return col;
    }
    return found;
  }, null);
}

const includeIn = columns => colKey => !!findColumn(colKey, columns);

const byColumnsOrder = columns => (a, b) => {
  const aIndex = columns.findIndex(({ key }) => key === a);
  const bIndex = columns.findIndex(({ key }) => key === b);
  return aIndex - bIndex;
};

class Table extends Component {
  static defaultProps = {
    hoverable: false,
    uniqueKey: 'id',
  };

  static propTypes = {
    /** enable hover style for row */
    hoverable: bool,
    /** columns controls */
    columns: arrayOf(
      shape({
        /** value of this key `props.data[key]` will be rendered into `<td/>` */
        key: string,
        /** for display as column title */
        label: string,
        /**
         * colum render function, if not provide
         * will simply render the value of that key
         *
         * @param {any} columnData value of `props.data[key]`
         * @param {object} allColumns `props.data`
         */
        render: func,
      })
    ).isRequired,
    /** table data */
    data: arrayOf(object).isRequired,
    /** the unique property (usually id) of individual object in `props.data` */
    uniqueKey: string,
  };

  renderRowCols(cols = {}) {
    const { columns } = this.props;
    return Object.keys(cols)
      .filter(includeIn(columns))
      .sort(byColumnsOrder(columns))
      .map(col => {
        const { render } = findColumn(col, columns);
        const hasRenderFunc = typeof render === 'function';
        const colData = cols[col];

        return (
          <td key={`llm-table-td-${col}`}>
            {hasRenderFunc ? render(colData, cols) : colData}
          </td>
        );
      });
  }

  renderRows(rows) {
    const { uniqueKey, hoverable } = this.props;
    return rows.map((row, index) => (
      <Row
        key={row[uniqueKey] || `llm-table-row-${index}`}
        hoverable={hoverable}
      >
        {this.renderRowCols(row)}
      </Row>
    ));
  }

  render() {
    // extract our own props so it doesn't pollute
    const { uniqueKey, data, columns, hoverable, ...remainProps } = this.props;

    return (
      <SCTable {...remainProps}>
        <thead>
          <tr>
            {columns.map(({ label, key }) => (
              <th key={`llm-table-th-${key}`}>{label}</th>
            ))}
          </tr>
        </thead>
        <tbody>{this.renderRows(data)}</tbody>
      </SCTable>
    );
  }
}

export default Table;
