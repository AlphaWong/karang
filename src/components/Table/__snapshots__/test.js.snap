// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshots 1`] = `
.c0 {
  width: 100%;
  border-spacing: 0;
  line-height: 1.4;
}

.c0 thead th {
  padding: 0.5em 1em;
  border-top: 1px rgba(229,229,229,1) solid;
  border-bottom: 1px rgba(229,229,229,1) solid;
  color: rgba(128,130,133,1);
  font-size: 12px;
  font-weight: 700;
  text-align: left;
}

.c0 tbody td {
  padding: 1em;
  vertical-align: top;
}

.c0 tbody td:nth-child(1) {
  border-left: 4px solid transparent;
}

.c1:nth-child(even) {
  background: #f9f9f9;
}

.c1:hover {
  background: #f2f2f2;
}

.c1:hover td:nth-child(1) {
  border-left-color: rgba(242,103,34,1);
}

<Table
  columns={
    Array [
      Object {
        "key": "name",
        "label": "Name",
      },
      Object {
        "key": "job",
        "label": "Occupation",
      },
      Object {
        "key": "hobbies",
        "label": "Fun",
      },
    ]
  }
  data={
    Array [
      Object {
        "hobbies": "basketball",
        "id": 123,
        "job": "CEO",
        "name": "Chan",
      },
      Object {
        "hobbies": "hockey",
        "id": 312,
        "job": "Hobo",
        "name": "Chow",
      },
      Object {
        "hobbies": "football",
        "id": 352,
        "job": "Carpenter",
        "name": "Mannings",
      },
      Object {
        "hobbies": "buy",
        "id": 24523,
        "job": "Business man",
        "name": "Wong",
      },
      Object {
        "hobbies": "eat",
        "id": 47657,
        "job": "Janitor",
        "name": "Smith",
      },
      Object {
        "hobbies": "sit",
        "id": 5675,
        "job": "PUTUS",
        "name": "Trump",
      },
    ]
  }
  hoverable={false}
  uniqueKey="id"
>
  <styled.table>
    <table
      className="c0"
    >
      <thead>
        <tr>
          <th
            key="llm-table-th-name"
          >
            Name
          </th>
          <th
            key="llm-table-th-job"
          >
            Occupation
          </th>
          <th
            key="llm-table-th-hobbies"
          >
            Fun
          </th>
        </tr>
      </thead>
      <tbody>
        <styled.tr
          hoverable={false}
          key="123"
        >
          <tr
            className="c1"
          >
            <td
              key="llm-table-td-name"
            >
              Chan
            </td>
            <td
              key="llm-table-td-job"
            >
              CEO
            </td>
            <td
              key="llm-table-td-hobbies"
            >
              basketball
            </td>
          </tr>
        </styled.tr>
        <styled.tr
          hoverable={false}
          key="312"
        >
          <tr
            className="c1"
          >
            <td
              key="llm-table-td-name"
            >
              Chow
            </td>
            <td
              key="llm-table-td-job"
            >
              Hobo
            </td>
            <td
              key="llm-table-td-hobbies"
            >
              hockey
            </td>
          </tr>
        </styled.tr>
        <styled.tr
          hoverable={false}
          key="352"
        >
          <tr
            className="c1"
          >
            <td
              key="llm-table-td-name"
            >
              Mannings
            </td>
            <td
              key="llm-table-td-job"
            >
              Carpenter
            </td>
            <td
              key="llm-table-td-hobbies"
            >
              football
            </td>
          </tr>
        </styled.tr>
        <styled.tr
          hoverable={false}
          key="24523"
        >
          <tr
            className="c1"
          >
            <td
              key="llm-table-td-name"
            >
              Wong
            </td>
            <td
              key="llm-table-td-job"
            >
              Business man
            </td>
            <td
              key="llm-table-td-hobbies"
            >
              buy
            </td>
          </tr>
        </styled.tr>
        <styled.tr
          hoverable={false}
          key="47657"
        >
          <tr
            className="c1"
          >
            <td
              key="llm-table-td-name"
            >
              Smith
            </td>
            <td
              key="llm-table-td-job"
            >
              Janitor
            </td>
            <td
              key="llm-table-td-hobbies"
            >
              eat
            </td>
          </tr>
        </styled.tr>
        <styled.tr
          hoverable={false}
          key="5675"
        >
          <tr
            className="c1"
          >
            <td
              key="llm-table-td-name"
            >
              Trump
            </td>
            <td
              key="llm-table-td-job"
            >
              PUTUS
            </td>
            <td
              key="llm-table-td-hobbies"
            >
              sit
            </td>
          </tr>
        </styled.tr>
      </tbody>
    </table>
  </styled.table>
</Table>
`;
